src/Microsoft.DotNet.Watcher.Tools/Internal/ProcessRunner.cs

```diff
@@ -0,0 +1,109 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Internal;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.DotNet.Watcher.Internal
+{
+    public class ProcessRunner
+    {
+        private readonly ILogger _logger;
+
+        public ProcessRunner(ILogger logger)
+        {
+            Ensure.NotNull(logger, nameof(logger));
+
+            _logger = logger;
+        }
+
+        public async Task<int> RunAsync(ProcessSpec processSpec, CancellationToken cancellationToken)
+        {
+            Ensure.NotNull(processSpec, nameof(processSpec));
+
+            var killLock = new object();
+            var tcs = new TaskCompletionSource<object>();
+            EventHandler handler = (sender, args) =>
+            {
+                tcs.TrySetResult(null);
+            };
+
+            var process = CreateProcess(processSpec);
+
+            process.Exited += handler;
+
+            var wasKilled = false;
+            var hasExited = false; // used to avoid race condition with process.Dispose()
```

Comment:
I don't see how it helps avoid a race condition - cancellation token fires while you are waiting in line 59. You go to line 47 and then the other thread resumes and the RunAsync runs to the end. The you are back on this thread and try to kill the process tree but the process has already been disposed. Also note that `wasKilled` flag in this scenario is also incorrect. 
I am not sure how helpful/interesting the `wasKilled` flag is but I would remove it - it makes this code even more complicated and I am not sure if it even is possible to get it always right. 
For the handler itself - I think it is fine to try to prevent killing the process if it has already exited but you have a couple non-atomic operations here some of them out of your control so you may just want to check if the process has not exited and then try to kill it if it hasn't and catch/swallow all the exceptions since there is not much you can do.